File: AI_UoB/programs/data_structures_for_ai/5_tree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BST:
    def __init__(self):
        self.root = None

    def insert(self, value):
        self.root = self._insert(self.root, value)

    def _insert(self, root, value):
        if root is None:
            return Node(value)
        if value < root.data:
            root.left = self._insert(root.left, value)
        elif value > root.data:
            root.right = self._insert(root.right, value)
        return root

    def search(self, value):
        return self._search(self.root, value)

    def _search(self, root, value):
        if root is None or root.data == value:
            return root
        if value < root.data:
            return self._search(root.left, value)
        else:
            return self._search(root.right, value)

    def findMinimum(self, root):
        while root.left is not None:
            root = root.left
        return root

    def delete(self, value):
        self.root = self._delete(self.root, value)

    def _delete(self, root, value):
        if root is None:
            return root
        if value < root.data:
            root.left = self._delete(root.left, value)
        elif value > root.data:
            root.right = self._delete(root.right, value)
        else:
            # Case 1 — no children
            if root.left is None and root.right is None:
                return None
            # Case 2 — one child
            elif root.left is None:
                return root.right
            elif root.right is None:
                return root.left
            # Case 3 — two children
            else:
                temp = self.findMinimum(root.right)
                root.data = temp.data
                root.right = self._delete(root.right, temp.data)
        return root

    def inorder(self):
        result = []
        self._inorder(self.root, result)
        print("Inorder:  ", result)

    def _inorder(self, root, result):
        if root is None:
            return
        self._inorder(root.left, result)
        result.append(root.data)
        self._inorder(root.right, result)

    def preorder(self):
        result = []
        self._preorder(self.root, result)
        print("Preorder: ", result)

    def _preorder(self, root, result):
        if root is None:
            return
        result.append(root.data)
        self._preorder(root.left, result)
        self._preorder(root.right, result)

    def postorder(self):
        result = []
        self._postorder(self.root, result)
        print("Postorder:", result)

    def _postorder(self, root, result):
        if root is None:
            return
        self._postorder(root.left, result)
        self._postorder(root.right, result)
        result.append(root.data)

if __name__ == "__main__":
    bst = BST()
    bst.insert(50)
    bst.insert(30)
    bst.insert(70)
    bst.insert(20)
    bst.insert(40)
    bst.insert(60)
    bst.insert(80)

    bst.inorder()
    bst.preorder()
    bst.postorder()

    found = bst.search(40)
    print(f"Search 40: {'Found' if found else 'Not Found'}")

    bst.delete(30)
    print("After deleting 30: ", end="")
    bst.inorder()

File: AI_UoB/programs/data_structures_for_ai/5_tree.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* left;
    struct Node* right;
};

struct Node* createNode(int value) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = value;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

struct Node* insert(struct Node* root, int value) {
    if (root == NULL)
        return createNode(value);
    if (value < root->data)
        root->left = insert(root->left, value);
    else if (value > root->data)
        root->right = insert(root->right, value);
    return root;
}

struct Node* search(struct Node* root, int value) {
    if (root == NULL || root->data == value)
        return root;
    if (value < root->data)
        return search(root->left, value);
    else
        return search(root->right, value);
}

struct Node* findMinimum(struct Node* root) {
    while (root->left != NULL)
        root = root->left;
    return root;
}

struct Node* delete(struct Node* root, int value) {
    if (root == NULL)
        return root;
    if (value < root->data)
        root->left = delete(root->left, value);
    else if (value > root->data)
        root->right = delete(root->right, value);
    else {
        // Case 1 — no children
        if (root->left == NULL && root->right == NULL) {
            free(root);
            return NULL;
        }
        // Case 2 — one child
        else if (root->left == NULL) {
            struct Node* temp = root->right;
            free(root);
            return temp;
        }
        else if (root->right == NULL) {
            struct Node* temp = root->left;
            free(root);
            return temp;
        }
        // Case 3 — two children
        else {
            struct Node* temp = findMinimum(root->right);
            root->data = temp->data;
            root->right = delete(root->right, temp->data);
        }
    }
    return root;
}

// Left → Root → Right
void inorder(struct Node* root) {
    if (root == NULL)
        return;
    inorder(root->left);
    printf("%d ", root->data);
    inorder(root->right);
}

// Root → Left → Right
void preorder(struct Node* root) {
    if (root == NULL)
        return;
    printf("%d ", root->data);
    preorder(root->left);
    preorder(root->right);
}

// Left → Right → Root
void postorder(struct Node* root) {
    if (root == NULL)
        return;
    postorder(root->left);
    postorder(root->right);
    printf("%d ", root->data);
}

int main() {
    struct Node* root = NULL;
    root = insert(root, 50);
    root = insert(root, 30);
    root = insert(root, 70);
    root = insert(root, 20);
    root = insert(root, 40);
    root = insert(root, 60);
    root = insert(root, 80);

    printf("Inorder:   "); inorder(root);   printf("\n");
    printf("Preorder:  "); preorder(root);  printf("\n");
    printf("Postorder: "); postorder(root); printf("\n");

    struct Node* found = search(root, 40);
    printf("Search 40: %s\n", found ? "Found" : "Not Found");

    root = delete(root, 30);
    printf("After deleting 30: "); inorder(root); printf("\n");

    return 0;
}